### PR TITLE
Health check

### PR DESCRIPTION
**Pull request by fboucquez**
Hi team, this is the bootstrap health check. It does everything defined in the original issue:

Fixes https://github.com/nemtech/symbol-bootstrap/issues/38

It checks:

- if containers are running
- if the opened ports are listening
- if gateways' /node/health are ok

This health check handles "repeat" and custom ports. 

This can be running with a `symbol-bootstrap run/start --healthCheck` or by calling its own command `symbol-bootstrap healthCheck`

Some other small changes:

- `ipv4_address` and custom `host` alias can be tunned per service (including broker). 
- A docker compose service can be excluded via `excludeDockerService`
- Improved how images are pulled. Better logging and images are prefetched so it doesn't affect compose up booting time 
- I've moved the custom preset examples to its own [file](https://github.com/nemtech/symbol-bootstrap/blob/health-check/docs/presetGuides.md) as I needed to add more. This is part of the doc refactoring @segfaultxavi . 

Fixes https://github.com/nemtech/symbol-bootstrap/issues/30

@Vektrat this will be the [guide](https://github.com/nemtech/symbol-bootstrap/blob/health-check/docs/presetGuides.md#disable-a-service) to disable a service

### CODE REVIEW

**Comment by segfaultxavi** [comment]
I think `docs/healthCheck.md` is missing from the PR.

**Comment by Vektrat** [comment]
thanks ^^

File: README.md [hunk 1]
```diff
@@ -75,142 +74,7 @@ Properties in each file override the previous values (by object deep merge).
...
--   Exclude a service via `repeat: 0`.
-
-If repeat is active, each value in the object is a [handlebars](https://handlebarsjs.com/guide/expressions.html) template. `$index` is the index of the generated service starting from 0. 
-
-Given an `$index` of 10:
-- `my-custom-peer-node-{{$index}}` will become `my-custom-peer-node-10`
-- `172.20.0.{{add $index 20}}` will become `172.20.0.30`
-
-The default preset `bootstrap` uses repeat but with just 1 database, 2 peers, 1 api and 1 rest gateway.
+Custom presets gives Symbol Bootstrap its versatility. Checkout the custom preset [guides](docs/presetGuides.md)!
```
Comment (segfaultxavi) [inline README.md:77]:
```suggestion
Custom presets give Symbol Bootstrap its versatility. Check out the custom preset [guides](docs/presetGuides.md)!
```

File: docs/healthCheck.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+`symbol-bootstrap healthCheck`
+==============================
+
+It checks if the services created with docker compose are up and running.
+
+* [`symbol-bootstrap healthCheck`](#symbol-bootstrap-healthcheck)
+
+## `symbol-bootstrap healthCheck`
+
+It checks if the services created with docker compose are up and running.
```
Comment (segfaultxavi) [inline docs/healthCheck.md:10]: I would list exactly what things it tests. There's more information in the issue description than in this README...

File: docs/run.md [hunk 3]
```diff
@@ -14,21 +14,24 @@ USAGE
...
 
-  -t, --timeout=timeout  [default: 60000] If running in detached mode, how long before timing out (in MS)
+  --healthCheck        If provided, the tool will review if the docker services are running correctly and report
 
-  --resetData            It reset the database and node data but keeps the generated configuration, keys, voting tree
-                         files and block 1
+  --resetData          It reset the database and node data but keeps the generated configuration, keys, voting tree
+                       files and block 1
+
+  --timeout=timeout    [default: 60000] If running in detached mode, how long before timing out (in MS)
```
Comment (segfaultxavi) [inline docs/run.md:34]:
```suggestion
  --timeout=timeout    [default: 60000] If running in detached mode, how long before timing out (in ms)
```
I guess you mean milliseconds, not megasiemens :)

File: docs/start.md [hunk 4]
```diff
@@ -29,23 +33,23 @@ OPTIONS
...
+                                          and report
+
   --report                                It generates reStructuredText (.rst) reports describing the configuration of
                                           each node.
 
   --resetData                             It reset the database and node data but keeps the generated configuration,
                                           keys, voting tree files and block 1
 
+  --timeout=timeout                       [default: 60000] If running in detached mode, how long before timing out (in
+                                          MS)
```
Comment (segfaultxavi) [inline docs/start.md:51]:
```suggestion
                                          ms)
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
Besides the small corrections I made, you're mixing different things in the same PR (Addition of `healthCheck` and moving of preset examples to their own README). But hey, it's your repo :)

File: docs/healthCheck.md (hunk 2 above)
Comment (fboucquez) [reply docs/healthCheck.md]: Fixed. The commands md files are generated from the TS command files.

**Comment by fboucquez** [comment]
> Besides the small corrections I made, you're mixing different things in the same PR (Addition of `healthCheck` and moving of preset examples to their own README). But hey, it's your repo :)

I've added more examples and split the file. The docs still need polish, they doc issue will still be open and yours.

File: docs/healthCheck.md [hunk 5]
```diff
@@ -1,21 +1,33 @@
 `symbol-bootstrap healthCheck`
 ==============================
 
-It checks if the services created with docker compose are up and running.
+It checks if the services created with docker compose are up and running:
+- if the docker containers are running.
+- if the services' exposed ports are listening.
+- if the rest gateways' /node/health are OK.
```
Comment (segfaultxavi) [inline docs/healthCheck.md:7]:
```suggestion
It checks if the services created with docker compose are up and running.

The performed checks are:
- If the docker containers are running.
- If the services' exposed ports are listening.
- If the rest gateways' /node/health are OK.
```

File: docs/healthCheck.md [hunk 6]
```diff
@@ -1,21 +1,33 @@
...
+
+The health check process handles 'repeat' and custom 'openPort' services.
 
 * [`symbol-bootstrap healthCheck`](#symbol-bootstrap-healthcheck)
 
 ## `symbol-bootstrap healthCheck`
 
-It checks if the services created with docker compose are up and running.
+It checks if the services created with docker compose are up and running:
```
Comment (segfaultxavi) [inline docs/healthCheck.md:15]: Colon at the end.

**Review by segfaultxavi — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -398,6 +262,7 @@ General users should install this tool like any other node module.
 * [`symbol-bootstrap clean`](docs/clean.md) - It removes the target folder deleting the generated configuration and data
 * [`symbol-bootstrap compose`](docs/compose.md) - It generates the `docker-compose.yml` file from the configured network.
 * [`symbol-bootstrap config`](docs/config.md) - Command used to set up the configuration files and the nemesis block for the current network
+* [`symbol-bootstrap healthCheck`](docs/healthCheck.md) - It checks if the services created with docker compose are up and running:
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply README.md]: The actual source is in the `src/commands/healthCheck.ts` file. The cli Oclif framework generates the doc.

File: docs/healthCheck.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+`symbol-bootstrap healthCheck`
+==============================
+
+It checks if the services created with docker compose are up and running.
+    
+- if the docker containers are running.
+- if the services' exposed ports are listening.
+- if the rest gateways' /node/health are OK.
```
Comment (segfaultxavi) [inline docs/healthCheck.md:8]:
```suggestion
The performed checks are:
- Whether the docker containers are running.
- Whether the services' exposed ports are listening.
- Whether the rest gateways' /node/health are OK.
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
One last minor fix!

Sorry for dragging this for so long, I think some of my previous comments have been lost because I remember I already pointed this one out.

File: docs/healthCheck.md (hunk 8 above)
Comment (fboucquez) [reply docs/healthCheck.md]: mmm, the "whether" wasn't in the old feedback. 

I cannot commit your change, healthCheck.md is generated. I've changed the typescript source code.

https://github.com/nemtech/symbol-bootstrap/blob/health-check/src/commands/healthCheck.ts#L7

Thanks!

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
Please be careful. The `if`->`Whether` change is not the only thing I requested.
There is an introductory sentence before the list of checks, otherwise the list does not make sense in the places where it appears on its own.

**Pull request merged**